### PR TITLE
Include support to `bytes`  built-in type.

### PR DESCRIPTION
**Pull request by pedropgusmao**
Implements support for `bytes` in response to #844 .

### CODE REVIEW

**Comment by Jasha10** [comment]
Hi @pedropgusmao, thanks for getting this started!
The implementation looks good.

Here are a few changes that I think should be made:
- `base.py:DictKeyType` should accept `bytes`.
- Update the OmegaConf docs

**Comment by Jasha10** [comment]
There are several places in OmegaConf's tests where each supported type is tested. To keep consistency with the other tests, it would be good to add `bytes` test cases in these places.
- `tests/test_omegaconf.py:test_is_none`
- `tests/test_omegaconf.py:test_is_interpolation`
- `tests/test_utils.py:test_node_wrap`
- `tests/test_utils.py:test_get_ref_type`
- `tests/test_utils.py:test_is_optional`
- `tests/test_nodes.py:test_accepts_mandatory_missing`
- `tests/test_nodes.py:test_legal_assignment`
- `tests/test_nodes.py:test_eq`
- `tests/test_pydev_resolver_plugin.py:test_get_dictionary_node``tests/test_pydev_resolver_plugin.py:test_can_provide`
- `tests/test_matrix.py:TestNodeTypesMatrix`
- `tests/__init__.py:SubscriptedDict`
- `tests/test_basic_ops_dict.py:TestDictKeyTypes`
- `tests/test_basic_ops_dict.py:test_in_dict`
- `tests/test_to_yaml.py:test_to_yaml`
- `tests/test_to_yaml.py:test_to_yaml_string_primitive_types_list``tests/test_to_yaml.py:test_to_yaml_string_primitive_types_dict`
- `tests/interpolation/test_interpolation.py:test_type_inherit_type`
- `tests/test_compare_dictconfig_vs_dict.py:TestUntypedDictConfig`
- `tests/test_compare_dictconfig_vs_dict.py:TestUntypeTestPrimitiveTypeDunderMethods`
- `tests/structured_conf`
- `tests/examples/test_dataclass_example.py:SimpleTypes`
- `tests/examples/test_dataclass_example.py:simple_types_class`
- comments in `tests/examples/test_dataclass_example.py`
- `tests/test_serialization.py:test_pickle_untyped`

**Comment by Jasha10** [comment]
An open questions:
Are there any values that OmegaConf's interpolation parser should interpret as bytes?
If so, we will need to modify the following:
- `omegaconf/grammar` and `omegaconf/grammar_parser.py`, etc
- `tests/test_grammar.py`
- `tests/interpolation/built_in_resolvers/test_oc_decode.py:test_decode`

File: omegaconf/nodes.py [hunk 1]
```diff
@@ -171,7 +171,11 @@ def _validate_and_convert_impl(self, value: Any) -> str:
 
         if OmegaConf.is_config(value) or is_primitive_container(value):
             raise ValidationError("Cannot convert '$VALUE_TYPE' to string: '$VALUE'")
-        return str(value)
+        if type(value) == bytes:
+            val = value.hex()
+        else:
+            val = value
+        return str(val)
```
Comment (Jasha10) [inline omegaconf/nodes.py:178]: This change is not obvious to me.
In vanilla python we have straightforward conversion of bytes to strings:
```python
b = b"abc123"
s = "abc123"
assert str(b) == s
```
Using the `bytes.hex` method here is counterintuitive to me. What is the motivation?

File: omegaconf/nodes.py [hunk 2]
```diff
@@ -218,6 +222,48 @@ def __deepcopy__(self, memo: Dict[int, Any]) -> "IntegerNode":
...
+            ),
+        )
+
+    def _validate_and_convert_impl(self, value: Any) -> bytes:
+        val: bytes
+        try:
+            if type(value) == bytes:
+                val = value
+            elif type(value) == str:
+                val = bytes.fromhex(value)
```
Comment (Jasha10) [inline omegaconf/nodes.py:252]: Similar to the above comment on use of the `bytes.hex` method -- using `bytes(value, "utf-8")` could be an alternative here.

**Comment by Jasha10** [comment]
I have pushed two commits to this PR branch.

**Comment by pedropgusmao** [comment]
Thanks @Jasha10 , 
I will start working on: https://github.com/omry/omegaconf/pull/845#issuecomment-999322636 (I might need help here though) and update the documentation for https://github.com/omry/omegaconf/pull/845#issuecomment-999322565 . 

Sorry, I'm not very familiar with OmegaConf's interpolation parser. Where could I learn more about it? https://github.com/omry/omegaconf/pull/845#issuecomment-999322685

File: omegaconf/nodes.py (hunk 2 above)
Comment (pedropgusmao) [reply omegaconf/nodes.py:252]: maybe it would be better to use `encode` and `decode` ?

**Comment by pedropgusmao** [comment]
working on tests:

- [ ] tests/test_omegaconf.py:test_is_none
- [ ] tests/test_omegaconf.py:test_is_interpolation
- [ ] tests/test_utils.py:test_node_wrap
- [ ] tests/test_utils.py:test_get_ref_type
- [ ] tests/test_utils.py:test_is_optional
- [ ] tests/test_nodes.py:test_accepts_mandatory_missing
- [ ] tests/test_nodes.py:test_legal_assignment
- [ ] tests/test_nodes.py:test_eq
- [ ] tests/test_pydev_resolver_plugin.py:test_get_dictionary_node``tests/test_pydev_resolver_plugin.py:test_can_provide
- [ ] tests/test_matrix.py:TestNodeTypesMatrix
- [ ] tests/__init__.py:SubscriptedDict
- [ ] tests/test_basic_ops_dict.py:TestDictKeyTypes
- [ ] tests/test_basic_ops_dict.py:test_in_dict
- [ ] tests/test_to_yaml.py:test_to_yaml
- [ ] tests/test_to_yaml.py:test_to_yaml_string_primitive_types_list``tests/test_to_yaml.py:test_to_yaml_string_primitive_types_dict
- [ ] tests/interpolation/test_interpolation.py:test_type_inherit_type
- [ ] tests/test_compare_dictconfig_vs_dict.py:TestUntypedDictConfig
- [ ] tests/test_compare_dictconfig_vs_dict.py:TestUntypeTestPrimitiveTypeDunderMethods
- [ ] tests/structured_conf
- [ ] tests/examples/test_dataclass_example.py:SimpleTypes
- [ ] tests/examples/test_dataclass_example.py:simple_types_class
- [ ] comments in tests/examples/test_dataclass_example.py
- [ ] tests/test_serialization.py:test_pickle_untyped

**Comment by Jasha10** [comment]
Don't worry about the parser for now -- we can publish a followup PR if we decide to change that later.
The idea is that arguments to omegaconf's [interpolations](https://omegaconf.readthedocs.io/en/latest/usage.html#variable-interpolation) and [resolvers](https://omegaconf.readthedocs.io/en/latest/custom_resolvers.html) get parsed according to a [grammar](https://omegaconf.readthedocs.io/en/latest/grammar.html#) that is defined in the files `omegaconf/grammar/*.g4`.
For example, the string `${oc.decode:abc}` would be parsed and decoded into a `str`, while `${oc.decode:123.456}` would be decoded into a `float`. My question above was about if there is anything that should be decoded into bytes... But this is a question that can be dealt with later.

File: omegaconf/nodes.py (hunk 2 above)
Comment (Jasha10) [reply omegaconf/nodes.py]: Nice!

File: tests/test_errors.py [hunk 3]
```diff
@@ -696,6 +708,18 @@ def finalize(self, cfg: Any) -> None:
         ),
         id="typed_DictConfig:assign_with_invalid_value,float",
     ),
+    param(
+        Expected(
+            create=lambda: DictConfig({"bar": BytesNode(b"abc123")}),
+            op=lambda cfg: cfg.__setattr__("bar", "x"),
+            exception_type=ValidationError,
+            msg="Value 'x' of type 'str' could not be converted to Bytes",
```
Comment (Jasha10) [inline tests/test_errors.py:716]:
```suggestion
            op=lambda cfg: cfg.__setattr__("bar", 123.4),
            exception_type=ValidationError,
            msg="Value '123.4' of type 'float' could not be converted to Bytes",
```
This should fix one of the failing tests.
Calling `__setattr__("bar", "x")` no longer raises a `ValidationError` because we are using the `str.encode` method now (thanks to commit https://github.com/omry/omegaconf/pull/845/commits/ec80f6ff433f0ec547cf9231e52fc3f11ddecd6f).

Comment (Jasha10) [reply tests/test_errors.py]: Resolved in https://github.com/omry/omegaconf/pull/845/commits/ee470da79e28057fe5b7ae72060088c30bd2babf

**Comment by pedropgusmao** [comment]
Hi @Jasha10 , I hope you are doing fine. 
Please take a look at the tests I've created so far and let me know if those need improving.
Also, when creating Bytes tests for `tests/test_to_yaml.py:test_to_yaml`, I have noticed that `pyyaml` doesn't like to use `bytes` as keys.  
```
>>> import yaml
>>> h = { "a" : "1", b"a": "1"}
>>> h
{'a': '1', b'a': '1'}
>>> print(yaml.dump(h))
a: '1'
? !!binary |
  YQ==
: '1'
```
Any ideas on how we should adapt this, or if `bytes` as keys are really necessary?

**Comment by pedropgusmao** [comment]
Hi @Jasha10 ,
Hope you are doing well and Happy New Year!
I was wondering if you had time to go over the tests I've written so far. 
Thanks!

**Comment by Jasha10** [comment]
Apologies for the late reply!
I will take a look at it shortly.
Happy New Year to you too :)

File: tests/test_utils.py [hunk 4]
```diff
@@ -692,14 +713,19 @@ def test_nullcontext() -> None:
         ),
         (
             lambda is_optional, missing: FloatNode(
-                value=10 if not missing else "???", is_optional=is_optional
+                value=10.0 if not missing else "???", is_optional=is_optional
```
Comment (pixelb) [inline tests/test_utils.py:716]: If this is a non related change, it would be better in a separate commit

Comment (pedropgusmao) [reply tests/test_utils.py]: Sorry about that. It was an unrelated change. It's been reverted now.

**Comment by Jasha10** [comment]
Looking good, @pedropgusmao!
After some consideration, @pixelb and I feel best about not doing any conversion between `str` and `bytes`. Not doing automatic conversion will help avoid surprising behavior and bugs.

**Comment by Jasha10** [comment]
FYI I'm planning to push a few more commits to round-out the testing.

File: tests/test_utils.py [hunk 5]
```diff
@@ -433,7 +458,7 @@ def test_type_str_none() -> None:
     [
         (Optional[int], "Optional[int]"),
         (Union[str, int, Color], "Union[str, int, Color]"),
-        (Optional[Union[int]], "Optional[int]"),
+        (Optional[int], "Optional[int]"),
```
Comment (Jasha10) [inline tests/test_utils.py:461]: I believe the use of `Optional[Union[int]]` here was intentional. Note the `Optional[int]` case above.

Comment (Jasha10) [reply tests/test_utils.py:461]: I'll push a commit to revert this.

File: tests/test_utils.py (hunk 4 above)
Comment (Jasha10) [reply tests/test_utils.py]: I've followed up in PR https://github.com/omry/omegaconf/pull/853.

Comment (Jasha10) [reply tests/test_utils.py]: I've followed up in PR https://github.com/omry/omegaconf/pull/853.

**Comment by Jasha10** [comment]
I've rebased to incorporate some changes from the `master` branch.

Additionally, I've added two commits.

The first commit (https://github.com/omry/omegaconf/pull/845/commits/32c0f9dbb6dfb4ed4585fc510de150cf26d4495f) removes all conversion between strings and bytes.
```python
from dataclasses import dataclass
from omegaconf import OmegaConf

@dataclass
class Conf:
    s: str = "abc"
    x: bytes = b"binary"

cfg = OmegaConf.create(Conf)

cfg.s = b"bin"  # previously performed conversion, now this throws ValidationError
cfg.x = "string"  # previously performed conversion, now this throws ValidationError
```

The second commit (https://github.com/omry/omegaconf/pull/845/commits/64eeeab208ebce0963609bea6b752f7a2ed9f2f0) adds more tests for `BytesNode` and polishes the tests and documentation.

**Comment by pedropgusmao** [comment]
Thanks for this, @Jasha10 
Is there anything else I can do on my side to help with this?
Also, just to clarify things a bit for me and to others who might read this PR; after we apply these changes, does it mean that :
- Calls to `instantiate`/ `call` that require `Bytes` to be passed as arguments will be allowed?
- `Bytes` will be supported, but not from `yaml` files. Or at least not encouraged?
- Conversions between `str` and `Bytes` will not be supported. 
Thanks

**Comment by Jasha10** [comment]
> Thanks for this, @Jasha10 Is there anything else I can do on my side to help with this?

Nope, I am happy with the shape of the PR.

---

> Also, just to clarify things a bit for me and to others who might read this PR; after we apply these changes, does it mean that :
> 
> * Calls to `instantiate`/ `call` that require `Bytes` to be passed as arguments will be allowed?

Yes, Hydra+instantiate will support binary (`bytes`) values.
```python
>>> instantiate({"_target_": "builtins.dict", "abc": b"123"})
{'abc': b'123'}
>>> instantiate({"_target_": "builtins.dict"}, abc=b"123")
{'abc': b'123'}
```
However, instantiate does not support binary keys:
```python
>>> instantiate({"_target_": "builtins.dict", b"binary_key": "foo"})
...
TypeError: keywords must be strings
```

---

> * `Bytes` will be supported, but not from `yaml` files.

Bytes will be supported in yaml files via yaml's [binary](https://yaml.org/type/binary.html) tag:
```python
>>> cfg = OmegaConf.create({"x": b"binary_data"})
>>> yml = OmegaConf.to_yaml(cfg)
>>> print(yml)
x: !!binary |
  YmluYXJ5X2RhdGE=

>>> OmegaConf.save(cfg, "x.yaml")
>>> print(open("x.yaml", "r").read())
x: !!binary |
  YmluYXJ5X2RhdGE=

>>> OmegaConf.load("x.yaml")
{'x': b'binary_data'}
```

This will work when Hydra loads yaml files too.

---

>  Or at least not encouraged?

Storing large binary data will not be encouraged.

The Hydra+OmegaConf stack is primarily intended as a human interface rather than as an arbitrary data store. In particular, OmegaConf and Hydra are not tuned to work with big data; they may make copies of data when performing oeprations such as config creation and merging.

Storing large blocks of binary data is better supported by something like `pickle` or `numpy.save`. Note that you can always use Hydra's instantiate API to invoke e.g. `numpy.load` when you need to get big data from disk. 
```python
big_data = instantiate({"_target_": "numpy.load", "file": "my_big_data.npy"})
```
For small blocks of binary data, there is `BytesNode` :)

---

> * Conversions between `str` and `Bytes` will not be supported.

Yes, automatic conversion will not be supported. Users will need to manually convert using python code or using a custom resolver.

>   Thanks

Thank you!!

**Comment by Jasha10** [comment]
I've force-pushed again after doing an interactive rebase to drop commit https://github.com/omry/omegaconf/commit/f4d4ec5477d6f88fdda6984a597544df1c4868cb from this PR. (The dropped commit would undo the diff from #853 which is now merged into `master`).

**Comment by Jasha10** [comment]
Whoops, I closed this PR on accident, and now it seems I'm unable to reopen it. I've opened PR 872 instead.
